### PR TITLE
Metadata: Add overloads for configuring shared type as owned type

### LINKED ISSUE #21892: Regression when having indexer properties on a generic entity
I have generic entities in my app and everything worked as expected before the regression. Adding < T > to User here and having indexer properties ( Data ) triggers the regression.

The last version it worked on is : 5.0.0-rc.1.20380.1

Exception:
```
The entity type 'System.Collections.Generic.Dictionary<string, object>' cannot be added to the model because a shared entity type with the same clr type already exists.
```

``` C#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Collections.Generic;

public class User<T>
{
    public int Id { get; set; }
    public Dictionary<string, object> Data { get; set; }
}

public class AppDbContext : DbContext
{
    public AppDbContext(DbContextOptions options) : base(options)
    {
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<User<int>>().OwnsOne(x => x.Data, x =>
        {
            x.IndexerProperty(typeof(string), "a").IsRequired(false);
            x.IndexerProperty(typeof(string), "b").IsRequired(false);
        });
    }
}

class Program
{
    static void Main(string[] args)
    {
        var sc = new ServiceCollection();
        var guid = Guid.NewGuid();
        sc.AddDbContext<AppDbContext>(op => op.UseSqlServer(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true;DataBase=" + guid));
        using (var provider = sc.BuildServiceProvider())
        {
            using (var scope = provider.CreateScope())
            using (var dbcontext = scope.ServiceProvider.GetService<AppDbContext>())
            {
                dbcontext.Database.EnsureCreated();
            }
        }

    }
}
```

EF Core version: 5.0.0-rc.1.20381.9
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 5.0
Operating system: Windows 10 Pro
IDE: Visual Studio 2019

### PR DESCRIPTION
**Pull request by smitpatel**
- Also fix issue in property mapping convention to exclude "shadow" properties on property bag which are no longer "shadow"

Resolves #21892

### CODE REVIEW

File: src/EFCore/Infrastructure/ModelValidator.cs [hunk 1]
```diff
@@ -151,7 +151,7 @@ public virtual void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.M
             {
                 var unmappedProperty = entityType.GetProperties().FirstOrDefault(
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ModelValidator.cs:152]:
```suggestion
                var unmappedProperty = entityType.GetDeclaredProperties().FirstOrDefault(
```

File: src/EFCore/Infrastructure/ModelValidator.cs [hunk 2]
```diff
@@ -151,7 +151,7 @@ public virtual void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.M
...
                     p => (!ConfigurationSource.Convention.Overrides(p.GetConfigurationSource())
-                            || !p.IsShadowProperty())
+                            || !(p.IsShadowProperty() || (p.DeclaringEntityType.IsPropertyBag && p.IsIndexerProperty())))
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ModelValidator.cs:154]: Add a note that this condition is only needed until https://github.com/dotnet/efcore/issues/14121 is implemented

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 3]
```diff
@@ -342,6 +375,7 @@ public virtual IndexBuilder HasIndex([NotNull] params string[] propertyNames)
             [NotNull] string navigationName)
             => OwnsOneBuilder(
                 new TypeIdentity(Check.NotNull(ownedType, nameof(ownedType)), (Model)Metadata.Model),
+                clrType: null,
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:378]: Add type to `TypeIdentity` and a property like `IsNamed` that returns `true` if a non-default name was specified

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 4]
```diff
@@ -376,7 +410,47 @@ public virtual IndexBuilder HasIndex([NotNull] params string[] propertyNames)
...
+        /// <returns> An object that can be used to configure the entity type. </returns>
+        public virtual EntityTypeBuilder OwnsOne(
+            [NotNull] string ownedTypeName,
+            [NotNull] Type ownedType,
+            [NotNull] string navigationName,
+            [NotNull] Action<OwnedNavigationBuilder> buildAction)
+        {
+            Check.NotEmpty(ownedTypeName, nameof(ownedTypeName));
+            Check.NotNull(ownedType, nameof(ownedType));
+            Check.NotNull(ownedType, nameof(ownedType));
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:449]: You really want to make sure it's not null 😆

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 5]
```diff
@@ -3011,7 +3014,21 @@ private static InternalIndexBuilder DetachIndex(Index indexToDetach)
...
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual InternalForeignKeyBuilder HasOwnership(
+            [NotNull] string targetEntityTypeName,
+            [NotNull] Type targetEntityType,
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs:3027]: Use `TypeIdentity` instead

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 6]
```diff
@@ -3091,7 +3111,14 @@ private static InternalIndexBuilder DetachIndex(Index indexToDetach)
                             }
                             else
                             {
-                                ModelBuilder.Entity(targetEntityType.Name, configurationSource, shouldBeOwned: true);
+                                if (targetClrType != null)
+                                {
+                                    ModelBuilder.SharedTypeEntity(targetEntityType.Name, targetClrType, configurationSource, shouldBeOwned: true);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs:3116]: `SharedTypeEntity` should throw if `model.HasEntityTypeWithDefiningNavigation(targetClrType)` returns `true`. All types with the same CLR type must be either shared or weak.

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:378]: There are places where we make certain decision based on typeIdentity.Type which would go wrong in those cases so needs to be updated. Which can make this much larger change.
Is it necessary? If yes, then can I submit it in a separate PR?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:378]: I think the places that need changing are fewer with `TypeIdentity` than without it (compared to current main), but you can do it in a separate PR.

Comment (smitpatel) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:378]: I will give it a try.

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 7]
```diff
@@ -523,7 +626,46 @@ private OwnedNavigationBuilder OwnsOneBuilder(in TypeIdentity ownedType, string
...
+            [NotNull] Type ownedType,
+            [NotNull] string navigationName,
+            [NotNull] Action<OwnedNavigationBuilder> buildAction)
+        {
+            Check.NotEmpty(ownedTypeName, nameof(ownedTypeName));
+            Check.NotNull(ownedType, nameof(ownedType));
+            Check.NotEmpty(navigationName, nameof(navigationName));
+            Check.NotNull(buildAction, nameof(buildAction));
+
+            buildAction(OwnsManyBuilder(new TypeIdentity(ownedTypeName), ownedType, navigationName));
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:668]:
```suggestion
            buildAction(OwnsManyBuilder(new TypeIdentity(ownedTypeName, ownedType), navigationName));
```

**Pull request merged**